### ISSUE TITLE
CESSDA Training Resources harvest

### ISSUE THREAD

**Issue by alen-vodopijevec-cessda** [SSHOC/sshoc-marketplace-backend#453]
CESSDA Training Resources on the MP are not up to date with current state on CESSDA side. Can you check the harvesting process?

**Comment by KlausIllmayer** [comment]
We would need to run a re-ingest. Currently, this needs some preparation. How urgent is it to have to the current state mirrored on the MP?

Adding @laureD19 @mkrzmr @olanowak to the discussion.

**Comment by alen-vodopijevec-cessda** [comment]
Resolving this during September would be fine. Is it possible?

**Comment by KlausIllmayer** [comment]
That could be tough, but I'm not able to give you a definitive answer. @olanowak is in charge of the ingest pipeline, maybe she can give us an answer here?

**Comment by alen-vodopijevec-cessda** [comment]
@olanowak

**Comment by helinek** [comment]
Hi, we were able to ingest CESSDA materials to the dev OM (https://sshoc-marketplace.acdh-dev.oeaw.ac.at/search?order=score&f.source=CESSDA). We are aware that there are a couple of duplicates but for now we are not able to tell the reason for that. We will have to investigate it further. As far as I know there should be 103 unique materials for CESSDA. @laureD19 Can you please check whether the ingested materials do not miss any information?

**Comment by alen-vodopijevec-cessda** [comment]
Hi! There are just 11 records show when filtered for CESSDA.

**Comment by KlausIllmayer** [comment]
That is due to bad timing. We had some issues with the development server and the re-index restarted half an hour ago. It will take some time (around 3-4 hours) until you will see all records.

**Comment by helinek** [comment]
We have actually reingested all of them...so now there are 115, I guess there are some duplicates again but as I mentioned we will have to investigate it.

**Comment by laureD19** [comment]
Hi all! Many thanks for the re-ingest @helinek - I'll take a look today or tomorrow. Just to be sure I understand how to review this ingest:
1. can you confirm that you re-aggregate using the existing pipeline? Did you change anything there? This endpoint - https://www.cessda.eu/sshoc-api - is still used as your source, correct?
2. I might understand it when checking the records, but can you explain how you handled the records that were modified in the MP between the first ingest and this re-ingest? I'm not sure we have cases on the dev instance, but we do have for sure such cases on production, so we need to agree on how to process them (without just overwriting the changes done in the MP).

see also, main issue about the CESSDA training ingest in the SSHOMP-Ingest repo: https://github.com/SSHOC/SSHOMP-Ingest/issues/25

@alen-vodopijevec-cessda - let us know too if you identify any missing/incorrect information

**Comment by helinek** [comment]
Hi @laureD19 

1. Yes, this is the endpoint that we used for ingestion. The reingestion was done from our local instance of DACE but in general we can confirm that it is an existing pipeline (we use the same endpoint and the same transformation rules as before, although I do not know when our previous ingestion took place)
2. This is actually something I do not know. I was not involved in ingesting the materials that you have on production. Ola was also not sure how those production materials were ingested - whether we did it somehow or it was done by someone else using different ingestion system. That's why I asked you to check whether there is nothing missing after our latest ingestion to dev. In terms of update - we try to update the items if they exist. However it might be the case that the backend API does remove some properties if they are not included in our update request (this is actually something that happened when I was ingesting Dariah.lab items with relations where I created relations with the dedicated endpoint and then I updated the whole item without having the relations included). We are of course open to make changes in our pipeline to handle modified records but we have to be sure that this is the case here for CESSDA materials.
I will also take a look to the issue you linked.

**Comment by KlausIllmayer** [comment]
Regarding the second point: @helinek it would be very important that you use the merge-API-calls. Otherwise changes done on the MP will be lost. We do have for every category such a merge endpoint, e.g. `POST /api/datasets/merge`. Can you confirm that you use this merge endpoint and also the merge workflow?

**Comment by helinek** [comment]
No, I cannot confirm that we use merge endpoints. I can see that we send a request to get the persistent id based on source id and source item id and if it's present we send PUT request for certain category. In case there is no persistent id for a source item then we just send POST request. I am quite new in SSHOC so I do not know the arrangements that were made for updating items. Can you please elaborate a bit on how we should use the MP API in order to update the exisiting items?

**Comment by KlausIllmayer** [comment]
This week I don't have much time to elaborate on this, but you can see a discussion and some pointers to the workflow in this old issue: https://github.com/SSHOC/sshoc-marketplace-backend/issues/97

In brief (without thinking too much about it):
1. collect all current items of a source in the MP
2. gather the data from the source one by one
3. if an item was already ingested (identified by source and sourceItemId - see the mentioned issue) do a merge (using the corresponding merge endpoint of the category - note: we don't have currently a workflow if the category changes at the source)
4. if an item was not found then create it (and use the correct sourceItemId)
5. in the end, if some items that were collected at point 1 are not covered in point 2, then it probably means that they were deleted at the source and should be set to deprecatedAtSource
6. moderators then take over and approve the items form the re-ingest - there could be, that the merge found conflicts which then must be solved by moderators (conflicts are handled by putting the version of MP and the version of source at the corresponding field, e.g. if title was changed at MP and was changed at source, merge will give title "Title at MP | Title at source" (or something similar)

Hope this helps - if more details are necessary, I can answer next week.
